### ISSUE TITLE
chimpraf.slr fails with error

### ISSUE THREAD

**Issue by janxkoci** [alanrogers/ape#1]
Hello,

I've managed to download the data, but when I try to convert them to raf using your script, it fails with the following error:

```bash
[koci@head ape]$ bash -x chimpraf.slr 
+ export LC_ALL=C
+ LC_ALL=C
+ fmt='panTro5/chr%d.hg19.panTro5.net.axt.gz\n'
+ seq 22
+ sort
+ xargs printf 'panTro5/chr%d.hg19.panTro5.net.axt.gz\n'
+ xargs zcat
+ axt2raf.py
+ sed 's/^chr//'
+ gzip -c
gzip: compressed data not written to a terminal. Use -f to force compression.
For help, type: gzip -h
Traceback (most recent call last):
  File "/home/users/koci/bin/axt2raf.py", line 281, in <module>
    a.pr()
  File "/home/users/koci/bin/axt2raf.py", line 176, in pr
    self.alt[i], self.raf[i])
IOError: [Errno 32] Broken pipe
xargs: zcat: terminated by signal 13
```

The same error occurs with the gorilla variant of the script. I'm not that good in python to figure out what's wrong. :confused: 

The `axt2raf.py` script is in `python2`, right? It's not obvious from the shebang (`#!/usr/bin/python`) but on our cluster this points to `python2`. Running it with `python3` gives syntax error.

**Comment by janxkoci** [comment]
**PS:** The same error occurs even if I just replace the script with this command:

```bash
zcat panTro5/*gz | axt2raf.py | sed 's/^chr//' | gzip -c
```

So it's not the slurm script, but the python tool.

**Comment by alanrogers** [comment]
This is just to let you know that I got your email and am looking into the problem.

Alan
________________________________

**Comment by alanrogers** [comment]
I'm going to ship the .raf files to your server so that I can take a little more time in debugging this.

Alan
________________________________

**Comment by alanrogers** [comment]
I just transferred two files to your server, clusterprf.osu.cz. They are in directory

  /home/users/data_sharing/rogers

One is the chigor.raf.gz file, and the other is the corresponding md5 checksum. To check that the
transfer worked, type

   md5sum *.md5sum

I'll transfer the other .raf.gz files this afternoon. Should I transfer the hg19 reference genome too?

Alan
________________________________

**Comment by alanrogers** [comment]
The only error messages we get are "broken pipe" and "xargs: zcat terminated by signal 13" (which also means broken pipe).

I tried piping the output of axt2raf.py into "head", and this generated the same error at the same line number:

$ zcat panTro5/chr22*gz | axt2raf.py  | head -n 1000 > /dev/null
Traceback (most recent call last):
  File "/uufs/chpc.utah.edu/common/home/u0028949/bin/axt2raf.py", line 281, in <module>
    a.pr()
  File "/uufs/chpc.utah.edu/common/home/u0028949/bin/axt2raf.py", line 176, in pr
    self.alt[i], self.raf[i])
IOError: [Errno 32] Broken pipe

I don't think there is anything wrong with the Python code. I ran the process to completion with no errors, piping the output to /dev/null. Your pipeline is shutting down for some reason, so that axt2raf.py is no longer able to write, and that is generating the error. 

 Suggestions:

1. Execute "zcat panTro5/chr22*gz | axt2raf.py", which will print output to the terminal. The top few lines of output should look like this:

    $ zcat panTro5/chr22*gz | axt2raf.py  | head 
    #chr	pos	ref	alt	raf
    chr22	16050001	g	.	1.000000
    chr22	16050002	a	.	1.000000
    chr22	16050003	t	.	1.000000
    chr22	16050004	c	.	1.000000
    chr22	16050005	t	.	1.000000

2. If that works, could it be that you are running out of time in your script?

**Comment by janxkoci** [comment]
Hello, thank you for the files.

I tried pipping into `head` and see the same output, except I see chr1 in `head`. Pipping to `less` shows full output, after quiting `less` I see the python error again. 

I'll now try saving the output to a file and then run `sed` and `gzip` on it. I think that should work.

Timeout shouldn't be an issue, I'm doing it interactively now (in tmux), not submitting it as a job.

Thank you again for your help. 
Jan

**Comment by alanrogers** [comment]
Good. That confirms that axt2raf.py is getting sensible input and producing sensible output, at least for the first few lines.

I specified chromosome 22 on the command line. Note: "zcat panTro5/chr22*gz". I suspect you had "zcat panTro5/*gz". Let me know if you figure out what's breaking that pipe.

Alan

**Comment by alanrogers** [comment]
Also let me know when you've got it working, so I can close this issue.

Alan

**Comment by alanrogers** [comment]
Your pipeline ends with "gzip -c", and doesn't redirect the output into a file. That may be the problem. The command would attempt to dump compressed data to the screen, which would be a mess. Perhaps the kernel is detecting this and killing gzip, which then breaks all pipes upstream of gzip.

This isn't a problem in the slurm script, because one of the slurm commands at the top redirects standard output into a file. But when you run this script in bash, the slurm commands aren't interpreted, so the output just goes to the screen.

Try putting " > chimp.raf.gz" at the end of your pipeline.

**Comment by janxkoci** [comment]
I think you nailed it here:

https://github.com/alanrogers/ape/blob/c02ad9131f1ec0ff2bcd6c5c96982056270b5d87/chimpraf.slr#L28-L34

I just used commands from the slurm script, without realizing there is output set there. Torque also has settable output, I just didn't think of that (and I was testing it interactively with bash anyway).

**PS:** I just tested the redirection of output and the error is gone.

I think we can close this.
